File: code/frontend/src/features/board/services/board-service.ts
import { apiClient } from '@/lib/api-client';
import type { ApiResponse } from '@/features/auth/types';
import type { TaskDto } from '@/features/task/types';
import type {
  BoardColumnDto,
  BoardDto,
  CreateColumnPayload,
  MoveTaskPayload,
  ReorderColumnsPayload,
  UpdateBoardSettingsPayload,
  UpdateColumnPayload,
} from '../types';

export const boardService = {
  getProjectBoard: async (projectId: string): Promise<BoardDto> => {
    const res = await apiClient.get<ApiResponse<BoardDto>>(`/projects/${projectId}/board`);
    return res.data.data;
  },

  getBoardById: async (boardId: string): Promise<BoardDto> => {
    const res = await apiClient.get<ApiResponse<BoardDto>>(`/boards/${boardId}`);
    return res.data.data;
  },

  updateBoardSettings: async (
    boardId: string,
    data: UpdateBoardSettingsPayload
  ): Promise<BoardDto> => {
    const res = await apiClient.patch<ApiResponse<BoardDto>>(`/boards/${boardId}/settings`, data);
    return res.data.data;
  },

  addColumn: async (boardId: string, data: CreateColumnPayload): Promise<BoardColumnDto> => {
    const res = await apiClient.post<ApiResponse<BoardColumnDto>>(`/boards/${boardId}/columns`, data);
    return res.data.data;
  },

  updateColumn: async (columnId: string, data: UpdateColumnPayload): Promise<BoardColumnDto> => {
    const res = await apiClient.patch<ApiResponse<BoardColumnDto>>(`/board-columns/${columnId}`, data);
    return res.data.data;
  },

  deleteColumn: async (columnId: string): Promise<void> => {
    await apiClient.delete<ApiResponse<void>>(`/board-columns/${columnId}`);
  },

  reorderColumns: async (boardId: string, data: ReorderColumnsPayload): Promise<BoardDto> => {
    const res = await apiClient.patch<ApiResponse<BoardDto>>(`/boards/${boardId}/columns/reorder`, data);
    return res.data.data;
  },

  moveTask: async (boardId: string, data: MoveTaskPayload): Promise<TaskDto> => {
    const res = await apiClient.post<ApiResponse<TaskDto>>(`/boards/${boardId}/tasks/move`, data);
    return res.data.data;
  },
};

File: code/backend/src/main/java/com/taskflow/modules/board/service/BoardService.java
package com.taskflow.modules.board.service;

import com.taskflow.modules.board.dto.BoardColumnDto;
import com.taskflow.modules.board.dto.BoardDto;
import com.taskflow.modules.board.dto.CreateBoardColumnRequest;
import com.taskflow.modules.board.dto.MoveTaskRequest;
import com.taskflow.modules.board.dto.ReorderColumnsRequest;
import com.taskflow.modules.board.dto.UpdateBoardColumnRequest;
import com.taskflow.modules.board.dto.UpdateBoardSettingsRequest;
import com.taskflow.modules.task.dto.TaskDto;

import java.util.UUID;

public interface BoardService {

    BoardDto getBoardByProjectId(UUID userId, UUID projectId);

    BoardDto getBoardById(UUID userId, UUID boardId);

    BoardDto updateBoardSettings(UUID userId, UUID boardId, UpdateBoardSettingsRequest request);

    BoardColumnDto addColumn(UUID userId, UUID boardId, CreateBoardColumnRequest request);

    BoardColumnDto updateColumn(UUID userId, UUID columnId, UpdateBoardColumnRequest request);

    void deleteColumn(UUID userId, UUID columnId);

    BoardDto reorderColumns(UUID userId, UUID boardId, ReorderColumnsRequest request);

    TaskDto moveTask(UUID userId, UUID boardId, MoveTaskRequest request);
}
